c_glib/arrow-dataset-glib/filter.h

```diff
@@ -0,0 +1,387 @@
+gad_in_expression_new(GADExpression *operand,
+                      GArrowArray *set);
+
+GArrowArray *
+gad_in_expression_get_set(GADInExpression *expression);
+
+/* arrow::dataset::CastExpression */
+
+#define GAD_TYPE_CAST_EXPRESSION (gad_cast_expression_get_type())
+G_DECLARE_DERIVABLE_TYPE(GADCastExpression,
+                         gad_cast_expression,
+                         GAD,
+                         CAST_EXPRESSION,
+                         GADUnaryExpression)
+struct _GADCastExpressionClass
+{
+  GADUnaryExpressionClass parent_class;
+};
+
+GADCastExpression *
+gad_cast_to_expression_new(GADExpression *operand,
+                           GArrowDataType *type,
+                           GArrowCastOptions *options);
+
+GADCastExpression *
+gad_cast_like_expression_new(GADExpression *operand,
+                             GADExpression *like,
+                             GArrowCastOptions *options);
+
+GArrowCastOptions *
+gad_cast_expression_get_options(GADCastExpression *expression);
+
+/* arrow::dataset::ScalarExpression */
+
+#define GAD_TYPE_EXPRESSION (gad_scalar_expression_get_type())
+G_DECLARE_DERIVABLE_TYPE(GADScalarExpression,
+                         gad_scalar_expression,
+                         GAD,
+                         SCALAR_EXPRESSION,
+                         GADExpression)
+struct _GADScalarExpressionClass
+{
+  GADExpressionClass parent_class;
+};
+
+/* TODO: GArrowScalar */
+GADScalarExpression *
+gad_scalar_expression_new(GArrowScalar *value);
+
+GArrowScalar *
+gad_scalar_expression_get_value(GADScalarExpression *expression);
+
+/* arrow::dataset::FieldExpression */
+
+#define GAD_TYPE_FIELD_EXPRESSION (gad_field_expression_get_type())
+G_DECLARE_DERIVABLE_TYPE(GADFieldExpression,
+                         gad_field_expression,
+                         GAD,
+                         FIELD_EXPRESSION,
+                         GADExpression)
+struct _GADFieldExpressionClass
+{
+  GADExpressionClass parent_class;
+};
+
+GADFieldExpression *
+gad_field_expression_new(const gchar *name);
+
+gchar *
+gad_field_expression_get_name(GADFieldExpression *expression);
+
+/* arrow::dataset::CustomExpression */
+
+#define GAD_TYPE_CUSTOM_EXPRESSION (gad_custom_expression_get_type())
+G_DECLARE_DERIVABLE_TYPE(GADCustomExpression,
+                         gad_custom_expression,
+                         GAD,
+                         CUSTOM_EXPRESSION,
+                         GADExpression)
+struct _GADCustomExpressionClass
+{
+  GADExpressionClass parent_class;
+};
+
+/* arrow::dataset::ExpressionEvaluator */
+
+#define GAD_TYPE_EXPRESSION_EVALUATOR (gad_expression_evaluator_get_type())
+G_DECLARE_DERIVABLE_TYPE(GADExpressionEvaluator,
+                         gad_expression_evaluator,
+                         GAD,
+                         EXPRESSION_EVALUATOR,
+                         GObject)
+struct _GAExpressionEvaluatorClass
+{
+  GObjectClass parent_class;
+};
+
+/* TODO: ArrowDatum */
```

Comment:
@kou Should we add `GArrowDatum`?

Reply:
@mrkn @kou The type an Expression will evaluate to can be derived using `Result<shared_ptr<DataType>> Expression::Validate(const Schema&)`

Reply:
You'll still need to handle scalar (in addition to array) evaluations. If it would help I can reopen https://issues.apache.org/jira/browse/ARROW-7151 after 0.17 is completed so that you always get an array

Reply:
Thanks for the information!

We don't bind `arrow::Datum` for now.